### PR TITLE
Add `&` and `&=` operators on integer types

### PR DESCRIPTION
**Pull request by MaxDesiatov**
These operators are required for StaticString flag checks.

### CODE REVIEW

**Review by compnerd — CHANGES_REQUESTED** [review]
Please apply them uniformly to the integer types.

File: Sources/Core/Int.swift [hunk 1]
```diff
@@ -8,22 +8,38 @@
 @frozen
 public struct Int {
   @usableFromInline
-  internal var _value: Builtin.Int32
+  internal var _value: Builtin.Word
```
Comment (compnerd) [inline Sources/Core/Int.swift:11]: While the change is correct, this is not related to the handling of `&` operations.

File: Sources/Core/UInt.swift [hunk 2]
```diff
@@ -8,13 +8,30 @@
 @frozen
 public struct UInt {
   @usableFromInline
-  internal var _value: Builtin.Int32
+  internal var _value: Builtin.Word
```
Comment (compnerd) [inline Sources/Core/UInt.swift:11]: Similar.

**Review by compnerd — CHANGES_REQUESTED** [review]

File: Sources/Core/UInt.swift [hunk 3]
```diff
@@ -9,6 +9,23 @@
...
+  public static func &= (lhs: inout UInt, rhs: UInt) {
+    lhs = UInt(Builtin.and_Int32(lhs._value, rhs._value))
+  }
+
+  @_transparent
+  public static func & (lhs: UInt, rhs: UInt) -> UInt {
+    var lhs = lhs
+    lhs &= rhs
+    return lhs
+  }
```
Comment (compnerd) [inline Sources/Core/UInt.swift:28]: Hmm, can we flip the implementation around?

```
public static func & (_ lhs: UInt, _ rhs: UInt) -> UInt {
  return UInt(Builtin.and_Int32(lhs._value, rhs._value))
}

public static func &= (_ lhs: inout UInt, _ rhs: UInt) {
  lhs = lhs & rhs
}
```

Comment (MaxDesiatov) [reply Sources/Core/UInt.swift:28]: The way it currently is in my PR is how "real" stdlib implemented it. I don't think that implementation had a clarifying comment, but I just assumed they had a good (maybe slight performance difference?) reason for it.

Comment (compnerd) [reply Sources/Core/UInt.swift:28]: Well, simply copying the reference implementation isn't particularly a goal :).

I'm open to the idea that we need to write this this way, but I would rather have some reason to do that.  I would rather that the implementation be clear and concise with good reason for the complexity.

I suspect that this will end up the same once the optimizer has run through - the `@_transparent` should fully inline the operation as if it were written there.  However, if I am incorrect and there is a size implication, then I would prefer the variant that has the smaller output (code) size.

Comment (MaxDesiatov) [reply Sources/Core/UInt.swift:28]: Are you referring to the potential code size difference in uSwift clients? If so, how should we measure that? Do we need some kind of test suite for code size testing?

Comment (compnerd) [reply Sources/Core/UInt.swift:25]: Yes, we really do need a way to track the code-size impact of changes.  I'd like to use bloaty for that, but I would be okay with anything which tracks the size over time.  I assume that for most embedded cases, we would want static linking, but, the size of the library either static or dynamic is important to consider.

Comment (MaxDesiatov) [reply Sources/Core/UInt.swift:25]: BTW, the actual stdlib does this for all numeric operators too: they use `Builtin` within assignment operators, and non-assignment operators are implemented in terms of those.

Comment (compnerd) [reply Sources/Core/UInt.swift:25]: Thats certainly good to know.  However, being 1:1 identical to the default standard library is a non-goal.  The desire is to make it easy to use as broad of a surface in space constrained environments as possible.  As a more concrete example, I don't expect `String` to be fully compatible because not all environments need to have all the various string encodings.  If we find that we can do anything to reduce code size (or improve clarity) I would favour those.

File: Sources/Core/Int32.swift [hunk 4]
```diff
@@ -6,6 +6,21 @@
...
   @usableFromInline
   internal var _value: Builtin.Int32
+
+  @_transparent
+  public init(_ _value: Builtin.Int32) {
+    self._value = _value
+  }
+
+  @_transparent
+  public static func &= (lhs: inout Int32, rhs: Int32) {
```
Comment (compnerd) [inline Sources/Core/Int32.swift:16]: Stylistic note: these parameters are not labeled when passed, so it is nicer to use `_ ...` here IMO.

**Review by compnerd — APPROVED** [review]
